### PR TITLE
ARROW-534 - WIP

### PR DESCRIPTION
**Pull request by tebeka**
Just for code review, not final code.

### CODE REVIEW

File: cpp/src/arrow/ipc/json-internal.cc [hunk 1]
```diff
@@ -957,6 +955,108 @@ class JsonArrayReader {
...
+
+    typename TypeTraits<T>::BuilderType builder(pool_, type);
+    for (int i = 0; i < length; ++i) {
+      if (!is_valid[i]) {
+        builder.AppendNull();
+        continue;
+      }
+
+      const rj::Value& val = json_data_arr[i];
+      // TODO: Can we have signed ints in Timestamp?
```
Comment (xhochy) [inline cpp/src/arrow/ipc/json-internal.cc]: Timestamps before 1970 are negative, so: yes.

File: integration/data/simple.json [hunk 2]
```diff
@@ -59,6 +81,30 @@
...
+          "name": "date_field",
+          "count": 5,
+          "VALIDITY": [1, 0, 0, 1, 1],
+          "DATA": [
+            0, 
+            19,
+            1883,
+            10300,
+            17217 
+          ]
```
Comment (wesm) [inline integration/data/simple.json]: Can you also add data generator fixtures to `integration_test.py`?

**Comment by wesm** [comment]
This will have some minor rebase conflicts with #347

**Comment by wesm** [comment]
Small rebase required

File: cpp/src/arrow/ipc/json-internal.cc (hunk 1 above)
Comment (tebeka) [inline cpp/src/arrow/ipc/json-internal.cc]: OK :)

File: integration/data/simple.json (hunk 2 above)
Comment (tebeka) [inline integration/data/simple.json]: OK

File: cpp/src/arrow/ipc/json-internal.cc [hunk 3]
```diff
@@ -940,6 +954,108 @@ class JsonArrayReader {
...
+      if (IsSignedInt<T>::value) {
+        DCHECK(val.IsInt() || val.IsInt64());
+        builder.Append(val.GetInt64());
```
Comment (wesm) [inline cpp/src/arrow/ipc/json-internal.cc]: How does rapidjson handle integers over 2^53? If it doesn't have any problems then this is OK

File: cpp/src/arrow/ipc/json-internal.cc [hunk 4]
```diff
@@ -940,6 +954,108 @@ class JsonArrayReader {
...
+      } else if (IsUnsignedInt<T>::value) {
```
Comment (wesm) [inline cpp/src/arrow/ipc/json-internal.cc]: Timestamps are always signed

File: cpp/src/arrow/ipc/json-internal.cc [hunk 5]
```diff
@@ -940,6 +954,108 @@ class JsonArrayReader {
...
+        DCHECK(val.IsUint() || val.IsUint64());
+        builder.Append(val.GetUint64());
+      } else {
+        // We are in the wrong function
+        return Status::Invalid(type->ToString());
+      }
```
Comment (wesm) [inline cpp/src/arrow/ipc/json-internal.cc]: Can you combine the logic for `TimeType` and `TimestampType` into a single function?

File: cpp/src/arrow/ipc/json-internal.cc [hunk 6]
```diff
@@ -940,6 +954,108 @@ class JsonArrayReader {
...
```
Comment (wesm) [inline cpp/src/arrow/ipc/json-internal.cc]: Can you use the same parsing code as for integers?

File: cpp/src/arrow/type.cc [hunk 7]
```diff
@@ -31,6 +31,7 @@ bool Field::Equals(const Field& other) const {
   return (this == &other) ||
          (this->name == other.name && this->nullable == other.nullable &&
              this->type->Equals(*other.type.get()));
+           //this->dictionary == other.dictionary && this->type->Equals(*other.type.get()));
```
Comment (wesm) [inline cpp/src/arrow/type.cc]: This can be deleted

File: cpp/src/arrow/type.h [hunk 8]
```diff
@@ -439,7 +439,15 @@ struct ARROW_EXPORT DateType : public FixedWidthType {
...
 
-enum class TimeUnit : char { SECOND = 0, MILLI = 1, MICRO = 2, NANO = 3 };
+enum class TimeUnit : char {
+  SECOND = 0,
+  MILLI = 1,
+  MICRO = 2,
+  NANO = 3,
+
+  MIN = SECOND,
+  MAX = NANO,
```
Comment (wesm) [inline cpp/src/arrow/type.h]: `TimeUnit` is a strongly typed enum, so these MIN/MAX fields aren't necessary. In Flatbuffers, the MIN/MAX values are these to help account for NULL.

File: cpp/src/arrow/ipc/metadata-internal.cc [hunk 9]
```diff
@@ -176,6 +214,38 @@ static Status UnionToFlatBuffer(FBB& fbb, const std::shared_ptr<DataType>& type,
...
+  return flatbuf::CreateTimestamp(fbb, TimeUnitToFlatbuffer(unit)).Union();
+}
+
+#define TIME_TO_FB(fbb, unit, type) \
+  flatbuf::Create ##  type(fbb, TimeUnitToFlatbuffer(unit)).Union();
+
+
+#define RETURN_BAD_TIME_UNIT(unit)                                  \
+  do {                                                              \
+        if ((unit < TimeUnit::MIN) || (unit > TimeUnit::MAX)) {     \
```
Comment (wesm) [inline cpp/src/arrow/ipc/metadata-internal.cc]: If `unit` is an invalid value, I don't believe it will have made it this far

**Review by wesm — COMMENTED** [review]
You need to add date and time test cases for https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/ipc-adapter-test.cc#L178 and https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/ipc-file-test.cc#L183. 

see e.g. https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/test-common.h#L283

Integration tests are blocked until ipc-file-test and ipc-adapter-test are passing , then we can add the data generators to integration_test.py. We'll also need to wait for ARROW-582 to get done

**Comment by wesm** [comment]
I note that a similar test fixture is missing in https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/ipc-json-test.cc

created https://issues.apache.org/jira/browse/ARROW-591

File: cpp/src/arrow/ipc/json-internal.cc (hunk 3 above)
Comment (tebeka) [inline cpp/src/arrow/ipc/json-internal.cc]: See so, the following code works:

```c++
#include <iostream>
#include <sstream>
#include "rapidjson/document.h"
#include "rapidjson/istreamwrapper.h"

using namespace rapidjson;

int
main() {

    std::stringstream in;
    in << "{";
    // 4611686018427387917
    in << "\"x\": " << ((uint64_t(1)<<62) + 13) << ",";
    in << "\"y\": 2";
    in << "}";
    IStreamWrapper isw(in);
    Document doc;

    doc.ParseStream(isw);
    uint64_t x = doc["x"].GetUint64();
    std::cout << "x = " << x << std::endl;
    uint64_t y = doc["y"].GetUint64();
    std::cout << "y = " << y << std::endl;
}
```

File: cpp/src/arrow/ipc/json-internal.cc (hunk 4 above)
Comment (tebeka) [inline cpp/src/arrow/ipc/json-internal.cc]: OK, fixing

File: cpp/src/arrow/ipc/json-internal.cc (hunk 5 above)
Comment (tebeka) [inline cpp/src/arrow/ipc/json-internal.cc]: Sure. WIll do

File: cpp/src/arrow/ipc/json-internal.cc (hunk 3 above)
Comment (wesm) [inline cpp/src/arrow/ipc/json-internal.cc]: Good to know. Suppose we'll have the proof in the working integration tests anyway

File: cpp/src/arrow/ipc/json-internal.cc (hunk 6 above)
Comment (tebeka) [inline cpp/src/arrow/ipc/json-internal.cc]: Good idea, will try to add `is_base_of` to the `ReadArray` method for `PrimitiveCType` and `BooleanType`.

File: cpp/src/arrow/ipc/metadata-internal.cc (hunk 9 above)
Comment (tebeka) [inline cpp/src/arrow/ipc/metadata-internal.cc]: Where will we fail? In the flatbuffer side of things?

Comment (wesm) [inline cpp/src/arrow/ipc/metadata-internal.cc]: Right, i.e. it should not be possible to return an invalid `TimeUnit` from a flatbuffer

**Comment by wesm** [comment]
@tebeka I opened https://issues.apache.org/jira/browse/ARROW-620 -- since you started on the JSON support in this patch, if you have time can you take this up in a new patch?

**Comment by tebeka** [comment]
@wesm OK, will look into [ARROW-620](https://issues.apache.org/jira/browse/ARROW-620)

**Pull request closed without merging**